torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,122 @@
+    return torch.clamp(tensor, TORCH_HALF_MIN, TORCH_HALF_MAX).half()
+
+def quntize_tensor(tensor, qtype):
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected tensor "
+            "to be a single tensor list  of type torch.Tensor."
+        )
+    if (qtype == QuantType.FP16):
+        return fp32_to_fp16_with_clamp(tensor)
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def quntize_tensor_list(tensor_list, qtype):
+    if not isinstance(tensor_list, list) or not all(
+        isinstance(p, torch.Tensor) for p in tensor_list
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be of type List[torch.Tensor]."
+        )
+    if (qtype == QuantType.FP16):
+        input_tensor = [fp32_to_fp16_with_clamp(t) for t in tensor_list]
+        return input_tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def dequantize_tensor(tensor, qtype, quant_loss=None):
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be a single tensor list  of type torch.Tensor."
+        )
+    if (qtype == QuantType.FP16):
+        if tensor.dtype != torch.float16:
+            raise RuntimeError(
+                "tensor dtype is {} while expected to be FP16.".format(
+                    tensor.dtype
+                )
+            )
+        elif tensor.dtype == torch.float16 and quant_loss is None:
+            return tensor.float()
+        else:
+            return tensor.float() / quant_loss
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def dequantize_tensor_list(output_tensor_list, qtype, quant_loss=None):
+    if not isinstance(output_tensor_list, list) or not all(
+        isinstance(p, torch.Tensor) for p in output_tensor_list
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be of type List[torch.Tensor]."
+        )
+    if (qtype == QuantType.FP16):
+        if output_tensor_list[0].dtype != torch.float16:
+            raise RuntimeError(
+                "tensor dtype is {} while expected to be FP16.".format(
+                    output_tensor_list.dtype
+                )
+            )
+        elif quant_loss is None:
+            tensor = [t.float() for t in output_tensor_list]
+            return tensor
+        elif quant_loss is not None:
+            tensor = [t.float() / quant_loss for t in output_tensor_list]
+            return tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    def wrapper(tensor, *args, **kwargs):
```

Comment:
Why is the first argument of the wrapper `tensor`? For both `all_gather` and `all_to_all` the first argument is a tensor list.